tools/dev-server.js

```diff
@@ -0,0 +1,158 @@
+const path = require("path");
+const { createServer } = require("http");
+const chokidar = require("chokidar");
+const karma = require("karma");
+const serve = require("serve-handler");
+const colors = require("colors");
+const boxen = require("boxen");
+const commandLineArgs = require("command-line-args");
+const serveConfig = require("../serve.json");
+const { Builder } = require("./builder");
+
+const KARMA_PORT = 9876;
+const SERVE_PORT = 5000;
+
+/** @type {import("command-line-args").OptionDefinition[]} */
+const optionList = [
+  {
+    name: "profile",
+    alias: "p",
+    description: "Name of profile to build.",
+    defaultValue: "w3c",
+    multiple: false,
+    type: String,
+  },
+  {
+    name: "interactive",
+    alias: "i",
+    description: "Run in interactive mode",
+    defaultValue: false,
+    type: Boolean,
+  },
+  {
+    name: "browsers",
+    description: "Browsers for Karma unit tests",
+    defaultOption: undefined,
+    multiple: true,
+    type: String,
+  },
+  {
+    name: "grep",
+    description: "Run specific tests using karma --grep",
+    defaultOption: undefined,
+    type: String,
+  },
+];
+
+let args;
+try {
+  args = commandLineArgs(optionList);
+} catch (err) {
+  console.error(colors.error(err.message));
+  process.exit(127);
+}
+
+const karmaConfig = karma.config.parseConfig(
+  path.join(__dirname, "../karma.conf.js"),
+  {
+    browsers: args.browsers,
+    autoWatch: false,
+    port: KARMA_PORT,
+    logLevel: karma.constants.LOG_WARN,
+    client: {
+      args: ["--grep", args.grep || ""],
+    },
+    mochaReporter: { ignoreSkipped: true },
+  }
+);
+const karmaServer = new karma.Server(karmaConfig);
+karmaServer.start();
+
+const devServer = createServer((req, res) => serve(req, res, serveConfig));
+devServer.listen(SERVE_PORT);
+devServer.on("error", onError);
+
+if (args.interactive) {
+  registerStdinHandler();
+} else {
+  const watcher = chokidar.watch("./src", { ignoreInitial: true });
+  watcher.on("all", onFileChange);
+  watcher.on("error", onError);
+}
+
+printWelcomeMessage(args);
+buildAndTest();
+
+function registerStdinHandler() {
+  // https://stackoverflow.com/a/12506613
+  const stdin = process.stdin;
+  stdin.setRawMode(true);
+  stdin.resume();
+  stdin.setEncoding("utf8");
+
+  stdin.on("data", async key => {
+    switch (key) {
+      case "\u0003":
+      case "q": {
+        return karma.stopper.stop(karmaConfig, code => {
+          setTimeout(() => process.exit(code), 0);
+        });
+      }
+      case "t":
+        return await buildAndTest();
+      case "h":
+        return printWelcomeMessage(args);
+      default:
+        process.stdout.write(key);
+    }
+  });
+}
+
+async function onFileChange(_event, _file) {
+  await buildAndTest();
```

Comment:
Could the karma.stopper also be used here?

Reply:
Not sure how. karma.stopper stops the server altogether, and based on my attempts, it fails to restart :confused: